npm/npm-debug-cli/dataplaneParser/parser/parser.go

```diff
@@ -0,0 +1,338 @@
+	iptableBuffer := bytes.NewBuffer(nil)
+	if len(filenames) > 0 {
+		byteArray, err := ioutil.ReadFile(filenames[0])
+
+		if err != nil {
+			fmt.Print(err)
+		}
+		for _, b := range byteArray {
+			iptableBuffer.WriteByte(b)
+		}
+		chains := p.parseIptablesChainObject(tableName, iptableBuffer.Bytes())
+		return iptable.NewIptables(tableName, chains)
+
+	} else {
+		cmdArgs := []string{"-t", string(tableName)}
+		cmd := exec.Command(util.IptablesSave, cmdArgs...)
+
+		cmd.Stdout = iptableBuffer
+		stderrBuffer := bytes.NewBuffer(nil)
+		cmd.Stderr = stderrBuffer
+
+		err := cmd.Run()
+
+		if err != nil {
+			_, err = stderrBuffer.WriteTo(iptableBuffer)
+			if err != nil {
+				panic(err)
+			}
+		}
+		chains := p.parseIptablesChainObject(tableName, iptableBuffer.Bytes())
+		return iptable.NewIptables(tableName, chains)
+
+	}
+}
+
+// parseIptablesChainObject create a map of iptable chain name and iptable chain object
+func (p *Parser) parseIptablesChainObject(tableName string, byteArray []byte) map[string]*iptable.IptablesChain {
+	chainMap := make(map[string]*iptable.IptablesChain)
+	tablePrefix := []byte("*" + tableName)
+	curReadIndex := 0
+	for curReadIndex < len(byteArray) {
+		line, nextReadIndex := p.parseLine(curReadIndex, byteArray)
+		curReadIndex = nextReadIndex
+		if bytes.HasPrefix(line, tablePrefix) {
+			break
+		}
+	}
+
+	for curReadIndex < len(byteArray) {
+		line, nextReadIndex := p.parseLine(curReadIndex, byteArray)
+		curReadIndex = nextReadIndex
+		if len(line) == 0 {
+			continue
+		}
+		if bytes.HasPrefix(line, util.CommitBytes) || line[0] == '*' {
+			break
+		} else if line[0] == ':' && len(line) > 1 {
+			// We assume that the <line> contains space - chain lines have 3 fields,
+			// space delimited. If there is no space, this line will panic.
+			spaceIndex := bytes.Index(line, util.SpaceBytes)
+			if spaceIndex == -1 {
+				panic(fmt.Sprintf("Unexpected chain line in iptables-save output: %v", string(line)))
+			}
+			chainName := string(line[1:spaceIndex])
+			if val, ok := chainMap[chainName]; ok {
+				val.SetData(line)
+				chainMap[chainName] = val
+			} else {
+				chainMap[chainName] = iptable.NewIptablesChain(chainName, line, make([]*iptable.IptablesRule, 0))
+			}
+		} else if line[0] == '-' && len(line) > 1 {
+			chainName, ruleStartIndex := p.parseChainNameFromRule(line)
+			val, ok := chainMap[chainName]
+			if !ok {
+				val = iptable.NewIptablesChain(chainName, []byte{}, make([]*iptable.IptablesRule, 0))
+			}
+			rules := val.Rules()
+			rules = append(rules, p.parseRuleFromLine(line[ruleStartIndex:]))
+			val.SetRules(rules)
+		}
+	}
+	return chainMap
+}
+
+// parseLine parse each line of the read-in byteArray of iptables-save
+func (p *Parser) parseLine(readIndex int, byteArray []byte) ([]byte, int) {
+	curReadIndex := readIndex
+
+	// consume left spaces
+	for curReadIndex < len(byteArray) {
+		if byteArray[curReadIndex] == ' ' {
+			curReadIndex++
+		} else {
+			break
+		}
+	}
```

Comment:
flip this around so your exit condition is first, and it reads better
```suggestion
	for curReadIndex < len(byteArray) {
		if byteArray[curReadIndex] != ' ' {
			break
		} 
		curReadIndex++
	}
```